streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java

```diff
@@ -695,6 +695,58 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
         EasyMock.verify(taskManager);
     }
 
+    @Test
+    public void shouldNotAccessTaskManagerWhenPendingShutdownInRunOnce() {
+        final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
+
+        internalTopologyBuilder.addSource(null, "source", null, null, null, topic1);
+
+        final StreamThread streamThread = createStreamThread(clientId, new StreamsConfig(configProps(true)), true);
+
+        final Map<TaskId, Set<TopicPartition>> activeTasks = new HashMap<>();
+        final List<TopicPartition> assignedPartitions = new ArrayList<>();
+
+        // assign single partition
+        assignedPartitions.add(t1p1);
+        assignedPartitions.add(t1p2);
+        activeTasks.put(task1, Collections.singleton(t1p1));
+        activeTasks.put(task2, Collections.singleton(t1p2));
+
+        streamThread.taskManager().setAssignmentMetadata(activeTasks, Collections.emptyMap());
+        final MockConsumer<byte[], byte[]> mockConsumer = (MockConsumer<byte[], byte[]>) streamThread.consumer;
+        final Map<TopicPartition, Long> beginOffsets = new HashMap<>();
+        beginOffsets.put(t1p1, 0L);
+        beginOffsets.put(t1p2, 0L);
+        mockConsumer.updateBeginningOffsets(beginOffsets);
+
+        final Thread callbackThread = new Thread(() -> {
```

Comment:
Why do we need a `Thread` here?

Reply:
We need a separate thread to trigger callbacks, otherwise the state won't proceed. Maybe there is a better way to do that?

Reply:
Can you elaborate? What do you mean by

> otherwise the state won't proceed